### PR TITLE
Support cookies with value `None`

### PR DESCRIPTION
**Pull request by tducret**
# Summary

There are 3 kinds of request cookie headers.
```py
request = httpx.Request("GET", "https://www.example.org", cookies=cookies)
request.headers["cookie"]  # <== This is the request cookie header we are talking about here
```

### `"name=value"`

This is the classic case, already supported through : 
```py
cookies.set(name="name", value="value", domain="example.org")
```

### `"name="`

Please note the trailing `=` sign here.
This is also supported, by passing an empty string as the cookie value :
```py
cookies.set(name="name", value="", domain="example.org")
```

### `"name"`

Please note there is **no** trailing `=` sign here.
This is not officially supported (yet), but can be achieved with :
```py
cookies.set(name="name", value=None, domain="example.org")
```

It works because the [cookie jar implementation](https://github.com/python/cpython/blob/0e4f73b8e457b1efa57b735205e8e85a3d11d9f2/Lib/http/cookiejar.py#L1335-L1336) in the standard library already deals with it.
Contrary to the `httpx.Cookies.set` method, the standard `Cookie` [doesn't define strict typing](https://github.com/python/cpython/blob/0e4f73b8e457b1efa57b735205e8e85a3d11d9f2/Lib/http/cookiejar.py#L761-L772) for the `value`.

So this PR changes the `httpx._models` in order to support a cookie `value` which is `None`.
Otherwise the typing is incorrect, forcing the `httpx` to add a `# type: ignore` instruction.

The getter functions had to be updated as well, as the `None` was considered as a not found value.

2 new test cases are added in this PR to ensure the proper behavior.

# Checklist

- [X] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [X] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

### CODE REVIEW

File: httpx/_models.py [hunk 1]
```diff
@@ -1008,7 +1008,7 @@ async def aclose(self) -> None:
                 await self.stream.aclose()
 
 
-class Cookies(typing.MutableMapping[str, str]):
+class Cookies(typing.MutableMapping[str, typing.Optional[str]]):
```
Comment (T-256) [inline httpx/_models.py:1011]:
```suggestion
class Cookies(typing.MutableMapping[str, str | None]):
```

**Review by T-256 — APPROVED** [review]

File: httpx/_models.py (hunk 1 above)
Comment (tducret) [reply httpx/_models.py:1011]: Hi @T-256 ! Actually, That was my initial attempt, but the test workflow [failed](https://github.com/encode/httpx/actions/runs/7914423735/job/21604123090) with this notation.

Comment (T-256) [reply httpx/_models.py:1011]: Interesting, no problem then...

**Comment by tomchristie** [comment]
Do `name=` and `name` have the same behavior?...

**Comment by tducret** [comment]
> Do name= and name have the same behavior?...

According to the [RFC for HTTP protocol](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-05#section-5.3) :

> If the name-value-pair string lacks a %x3D ("=") character, then the name string is empty, and the value string is the value of name-value-pair.

So it may depend on the server implementation, but I can confirm a case where the server considered `name=` and `name` as different.
Although this was already possible to create both kinds of cookies, this PR confirms this usage by changing the type hint.

File: httpx/_models.py [hunk 2]
```diff
@@ -1080,23 +1082,26 @@ def get(  # type: ignore
         default: str | None = None,
         domain: str | None = None,
         path: str | None = None,
+        raise_when_not_found: bool = False,
```
Comment (tomchristie) [inline httpx/_models.py:1085]: Could we avoid adding this API as part of the pull request?

File: tests/models/test_cookies.py [hunk 3]
```diff
@@ -30,6 +30,16 @@ def test_cookies_update():
...
 
 
+def test_cookie_value_can_be_none():
+    cookies = httpx.Cookies()
+    more_cookies = httpx.Cookies()
+    more_cookies.set("no-value", None, domain="example.com")
+
+    cookies.update(more_cookies)
+    assert dict(cookies) == {"no-value": None}
+    assert cookies.get("no-value", domain="example.com") is None
```
Comment (tomchristie) [inline tests/models/test_cookies.py:40]: It's not absolutely clear to me that allowing `None` here is an API improvement. 🤔

File: httpx/_models.py (hunk 2 above)
Comment (tducret) [reply httpx/_models.py:1085]: Here, I wanted to preserve this responsibility of raising the `KeyError` exception to the `__getitem__()` method.

However, if it's safe to raise it whenever it's not found through `get()`, then we could remove `raise_when_not_found` parameter.

What do you think?

File: tests/models/test_cookies.py (hunk 3 above)
Comment (tducret) [reply tests/models/test_cookies.py]: I changed the test to `assert cookies["no-value"] is None`.
I also added `test_get_missing_cookie_raises_exception` to highlight the difference between getting a missing cookie (raises exception) and getting a cookie whose value is `None`.

Is it better?

**Comment by tomchristie** [comment]
Thanks @tducret. Appreciate your time. I think we'll pass on this.
I think we should just treat `name` and `name=` identically.
Ie. both should be the empty string.

**Comment by tducret** [comment]
Alright @tomchristie, no worries.

**Comment by tomchristie** [comment]
(Apologies for the delay)

**Pull request closed without merging**